### PR TITLE
Katacoda Thanos scenarios: Global View course

### PR DESCRIPTION
**Pull request by bwplotka**
Preview: https://katacoda.com/adrien-f

Making interactive tutorials for Thanos happen with @adrien-f and Katacoda :heart:

Feedback welcome. So far we are planning 5 scenarios:

```
  {
      "course_id": "globalview",
      "title": "Intro: Globalview and seemless HA for Prometheus",
      "description": "Learn how to easily transform Prometheus into centralized, highly available monitoring using Thanos."
    },
    {
      "course_id": "lts",
      "title": "Downsampling and unlimited metric retention with Prometheus using Thanos",
      "description": "Learn how to extend you metric retention in a cheap way."
    },
    {
      "course_id": "meta-monitoring",
      "title": "Global and meta alerts with Thanos Ruler",
      "description": "TBD"
    },
    {
      "course_id": "cross-cluster-comm",
      "title": "Connecting remote Prometheuses to Thanos using simple Envoy setup.",
      "description": "TBD"
    },
    {
      "course_id": "caching",
      "title": "Query low tail latency with low cost: Introducing caching to Thanos",
      "description": "TBD"
    }
```

### CODE REVIEW

**Comment by bwplotka** [comment]
The first course is ready, so I think I would be happy to merge this (:

We can iterate to add more later.

**Comment by adrien-f** [comment]
Looking over my lunch break!

File: docs/getting-started.md [hunk 1]
```diff
@@ -234,12 +265,8 @@ _NOTE: The compactor must be run as a **singleton** and must not run when manual
...
 which does rule and alert evaluation on top of a given Thanos Querier.
 
-### Receiver
-
-TBD
-
 ## Extras
 
-Thanos also has a tutorial on deploying it to Kubernetes. We have a full page describing a standard deployment here.
+Thanos also has a tutorial on deploying it to Kubernetes.
```
Comment (adrien-f) [inline docs/getting-started.md]: No link ?

File: tutorials/katacoda/thanos/1-globalview/intro.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
+Welcome to the Thanos introduction!
+
+[Thanos](https://thanos.io) is a set of components that can be composed into a highly available metric system with unlimited storage capacity.
+It can be added seamlessly on top of existing Prometheus deployments.
+
+Thanos provides a global query view, data backup, and historical data access as its core features. 
+All three features can be run independently of each other. This allows you to have a subset of Thanos features ready for 
+immediate benefit or testing, while also making it flexible for gradual adoption in more complex environments. 
+
+Thanos will work in cloud native environments like Kubernetes as well as more traditional ones. However this course uses docker
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/intro.md]:
```suggestion
Thanos will work in cloud native environments like Kubernetes as well as more traditional ones. However, this course uses docker 
```

File: tutorials/katacoda/thanos/1-globalview/step1.md [hunk 3]
```diff
@@ -0,0 +1,173 @@
+# Step 1 - Start initial Prometheus servers
+
+Thanos is meant to scale and extend vanilla Prometheus. This means that you can gradually, without disruption, deploy Thanos on top of your existing Prometheus setup.
+
+Let's start our tutorial by spinning up three Prometheus servers. Why three?
+The real advantage of Thanos is when you need to scale out Prometheus from single replica. Some reason of scale out might be:
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/step1.md]:
```suggestion
The real advantage of Thanos is when you need to scale out Prometheus from a single replica. Some reason for scale-out might be:
```

File: tutorials/katacoda/thanos/1-globalview/step1.md [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+* Adding functional sharding because of metrics high cardinality 
+* Need for high availability of Prometheus e.g: Rolling upgrades
+* Aggregating queries from multiple clusters
+
+For the purpose of the this course, let's imagine following situation:
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/step1.md]:
```suggestion
For this course, let's imagine the following situation:
```

File: tutorials/katacoda/thanos/1-globalview/step1.md [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    --name prometheus-0-eu1 \
+    quay.io/thanos/prometheus:v2.12.0-rc.0-rr-streaming \
+    --config.file=/etc/prometheus/prometheus.yml \
+    --storage.tsdb.path=/prometheus \
+    --web.listen-address=:9090 \
+    --web.enable-lifecycle \
+    --web.enable-admin-api && echo "Prometheus EU1 started!"
+```{{execute}}
+
+NOTE: We are using latest Prometheus image for latest remote read protocol.
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/step1.md]:
```suggestion
NOTE: We are using the latest Prometheus image so we can take profit from the latest remote read protocol. 
```

File: tutorials/katacoda/thanos/1-globalview/step1.md [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+Why would one need multiple Prometheus instances?
+
+* High Availability (multiple replicas)
+* Scaling ingestion: Functional Sharding
+* Multi cluster/environment architecture  
+
+## Problem statement: Global view challenge
+
+Let's try to play with this setup a bit. You are free to query any metrics, however let's try to fetch some certain information from
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/step1.md]:
```suggestion
Let's try to play with this setup a bit. You are free to query any metrics, however, let's try to fetch some certain information from 
```

File: tutorials/katacoda/thanos/1-globalview/step1.md [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+Try to get this information from the current setup!
+
+To see the answer to this question click SHOW SOLUTION below. 
+
+## Next
+
+Great! We have now running 3 Prometheus instances.
+
+In the next steps we will learn how we can install Thanos on top of our initial Prometheus setup to solve problems shown in the challenge.
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/step1.md]:
```suggestion
In the next steps, we will learn how we can install Thanos on top of our initial Prometheus setup to solve problems shown in the challenge.
```

File: tutorials/katacoda/thanos/1-globalview/step2.md [hunk 8]
```diff
@@ -0,0 +1,163 @@
+# Step 2 - Installing Thanos sidecar
+
+Let's take the setup from previous step and seamlessly install Thanos to add Global View with HA handling feature.
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/step2.md]:
```suggestion
Let's take the setup from the previous step and seamlessly install Thanos to add Global View with HA handling feature.
```

File: tutorials/katacoda/thanos/1-globalview/step2.md [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+
+## Thanos Components 
+
+Thanos is a single Go binary capable to run in different modes. Each mode represents different
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/step2.md]:
```suggestion
Thanos is a single Go binary capable to run in different modes. Each mode represents a different
```

File: tutorials/katacoda/thanos/1-globalview/step2.md [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+component and can be invoked in single command.
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/step2.md]:
```suggestion
component and can be invoked in a single command.
```

File: tutorials/katacoda/thanos/1-globalview/step2.md [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+
+Let's take a look on all the Thanos commands:
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/step2.md]:
```suggestion
Let's take a look at all the Thanos commands: 
```

File: tutorials/katacoda/thanos/1-globalview/step2.md [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+In this step we will focus on `thanos sidecar`:
+
+```
+  sidecar [<flags>]
+    sidecar for Prometheus server
+```
+
+## Sidecar
+
+Sidecar as name suggests should be deployed together with Prometheus. Sidecar has multiple features:
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/step2.md]:
```suggestion
Sidecar as the name suggests should be deployed together with Prometheus. Sidecar has multiple features:
```

File: tutorials/katacoda/thanos/1-globalview/step2.md [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+* It is essentially in further long term storage options described in [next]() courses.
+* It is capable to watch for configuration and Prometheus rules (alerting or recording) and notify Prometheus for dynamic reloads:
+  * optionally substitute with environment variables
+  * optionally decompress if gzipp-ed
+   
+You can read more about sidecar [here](https://thanos.io/components/sidecar.md/)
+
+## Installation
+
+To allow Thanos to efficiently query Prometheus data, let's install sidecar to each Prometheus instances we deployed in previous step as shown below:
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/step2.md]:
```suggestion
To allow Thanos to efficiently query Prometheus data, let's install sidecar to each Prometheus instances we deployed in the previous step as shown below:
```

File: tutorials/katacoda/thanos/1-globalview/step2.md [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+    --prometheus.url http://127.0.0.1:9092 && echo "Started sidecar for Prometheus 1 US1"
+```{{execute}}
+
+## Verification
+
+Now, to check if sidecars are running well, let's modify Prometheus scrape configuration to include our added sidecars.
+ 
+As always click `Copy To Editor` for each config to propagate the configs to each file. 
+
+Note that only thanks to the sidecar, all those changes will be immdiately reloaded and updated in Prometheus!
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/step2.md]:
```suggestion
Note that only thanks to the sidecar, all those changes will be immediately reloaded and updated in Prometheus!
```

File: tutorials/katacoda/thanos/1-globalview/step2.md [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+
+Now you should see new, updated configuration on each Prometheus. For example here in [Prometheus 0 EU1 /config](https://[[HOST_SUBDOMAIN]]-9090-[[KATACODA_HOST]].environments.katacoda.com/config).
+In the same time [`up`](https://[[HOST_SUBDOMAIN]]-9090-[[KATACODA_HOST]].environments.katacoda.com/graph?g0.expr=up&g0.tab=1) should show `job=sidecar` metrics. 
+
+Since now Prometheus has access to sidecar metrics we can query for [`thanos_sidecar_prometheus_up`](https://[[HOST_SUBDOMAIN]]-9090-[[KATACODA_HOST]].environments.katacoda.com/graph?g0.expr=thanos_sidecar_prometheus_up&g0.tab=1)
+to check if sidecar has access to Prometheus.
+
+## Next 
+
+Great! Now you should have setup deployed as in presented image:
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/step2.md]:
```suggestion
Great! Now you should have setup deployed as in the presented image:
```

File: tutorials/katacoda/thanos/1-globalview/step2.md [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+
+![sidecar install](https://docs.google.com/drawings/d/e/2PACX-1vRHlEJd9OVH80hczxkqZKYDVXxwugX55VWKtLJhS6R7D3BbmkBW9qGyyD4JyLbAe9CK9EzvurWTagTR/pub?w=1058&h=330)
+
+In the next step, we will add a final component allowing us to fetch Prometheus metrics from the very single endpoint.
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/step2.md]:
```suggestion
In the next step, we will add a final component allowing us to fetch Prometheus metrics from a single endpoint.
```

File: tutorials/katacoda/thanos/1-globalview/step3.md [hunk 17]
```diff
@@ -0,0 +1,124 @@
+# Step 3 - Adding Thanos Querier
+
+Thanks to the previous step we have three running Prometheus instances with a sidecar each. In this step we will install 
+Thanos Querier which will use sidecars and allow querying all metrics from the single place as presented below:
+
+![with querier](https://docs.google.com/drawings/d/e/2PACX-1vSB9gN82px0lxk9vN6wNw3eXr8Z0EVROW3xubsq7tgjbx_nXsoZ02ElzvxeDevyjGPWv-f9Gie0NeNz/pub?w=926&h=539)
+
+But before that, let's look closer what Querier component does:
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/step3.md]:
```suggestion
But before that, let's take a closer look at what the Querier component does:
```

File: tutorials/katacoda/thanos/1-globalview/step3.md [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+
+## Querier
+
+The Querier component (also called "Query") is essentially a vanilla PromQL Prometheus engine that fetches the data from any service
+that implements Thanos [StoreAPI](https://thanos.io/integrations.md/#storeapi). This means that Querier exposes the Prometheus HTTP v1 API to query the data in a common PromQL language.
+This allows compatibility with Grafana or other consumers of Prometheus' API.
+
+Additionally Querier is capable to deduplicate StoreAPIs that are in the same HA group. We will see how it
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/step3.md]:
```suggestion
Additionally, Querier is capable of deduplicating StoreAPIs that are in the same HA group. We will see how it 
```

File: tutorials/katacoda/thanos/1-globalview/step3.md [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+
+## Setup verification
+
+Thanos Querier exposes very similar UI to the Prometheus, but on top of many `StoreAPIs you wish to connect to.
+
+To check if the Querier works as intended let's look on [Querier UI `Store` page](https://[[HOST_SUBDOMAIN]]-29090-[[KATACODA_HOST]].environments.katacoda.com/stores). 
+
+This should list all our three sidecars, including their external labels.
+  
+## Global view - Not challeging anymore?
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/step3.md]:
```suggestion
## Global view - Not challenging anymore?
```

File: tutorials/katacoda/thanos/1-globalview/step3.md [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+
+Now, let's get back to our challenge from step 1, so finding the answer to  **How many series (metrics) we collect overall on all Prometheus instances we have?**  
+
+With the querier this is now super simple. 
+
+It's just enough to query Querier for <a href="https://[[HOST_SUBDOMAIN]]-29090-[[KATACODA_HOST]].environments.katacoda.com/graph?g0.range_input=1h&g0.expr=sum(prometheus_tsdb_head_series)&g0.tab=1&g1.range_input=5m&g1.expr=prometheus_tsdb_head_series&g1.tab=0">`sum(prometheus_tsdb_head_series)`</a>
+
+You should see the single value representing number of series scraped in both clusters in the current mode.
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/step3.md]:
```suggestion
You should see the single value representing the number of series scraped in both clusters in the current mode.
```

File: tutorials/katacoda/thanos/1-globalview/step3.md [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+prometheus_tsdb_head_series{cluster="us1",instance="127.0.0.1:9092",job="prometheus",replica="1"}
+```
+
+So how Thanos Querier knows how to deduplicate correctly?
+
+If we would look again into Querier configuration we can see that we also set `quey.replica-label` flag. 
+This is exactly the label Querier will try to deduplicate by for HA groups. This means that any metric with exactly
+the same labels *except replica label* will be assumed as the metric from the same HA group, and deduplicated accordingly.
+
+If we would open `prometheus1_us1.yml` config file in editor or if you go to Prometheus 1 US1 [/config](https://[[HOST_SUBDOMAIN]]-9090-[[KATACODA_HOST]].environments.katacoda.com/config).
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/step3.md]:
```suggestion
If we would open `prometheus1_us1.yml` config file in the editor or if you go to Prometheus 1 US1 [/config](https://[[HOST_SUBDOMAIN]]-9090-[[KATACODA_HOST]].environments.katacoda.com/config).
```

File: tutorials/katacoda/thanos/1-globalview/step3.md [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+Once by `replica=1, cluster=us1` Prometheus and once by `replica=0, cluster=us1` Prometheus. If we configure Querier to
+deduplicate by `replica` we can transparently handle this High Available pair of Prometheus instances to the user.
+
+## Production deployment
+
+Normally Querier runs in some central global location (e.g next to Grafana) with remote access to all Prometheus-es (e.g via ingress, proxies vpn or peering)
+
+You can also stack (federate) Queriers on top of other Queries, as Query expose `StoreAPI` as well!
+
+More information about those advanced topics can be found in next courses that will be added soon.
```
Comment (adrien-f) [inline tutorials/katacoda/thanos/1-globalview/step3.md]:
```suggestion
More information about those advanced topics can be found in the next courses that will be added soon.
```

File: go.mod [hunk 23]
```diff
@@ -63,3 +63,5 @@ replace (
 	k8s.io/klog => k8s.io/klog v0.3.1
 	k8s.io/kube-openapi => k8s.io/kube-openapi v0.0.0-20190228160746-b3a7cee44a30
 )
+
+go 1.13
```
Comment (bwplotka) [inline go.mod]: will remove this

**Pull request merged**